src/frontend/src/utils/countdown.ts

```diff
@@ -132,14 +132,22 @@ export class AsyncCountdown {
   // XXX: Some assumptions are made in the code that the countdown is never restarted.
   private stopped: boolean;
 
+  // A promise that resolves once the countdown has stopped
+  private promise: Promise<void>;
+  private resolve?: () => void;
+
   // when it should stop (seconds since epoch)
   constructor(private expirationSeconds: number) {
     this.stopped = this.pastExpiration();
+    this.promise = new Promise((resolve) => {
+      this.resolve = resolve;
+    });
   }
 
   // Stop the countdown explicitly
   stop() {
     this.stopped = true;
+    this.resolve?.();
```

Comment:
idea: would it work / be interesting to set `stopped` to true when it's resolved or both are not related?

```
(this.resolve ?? Promise.resolve()).then(() => this.stopped = true));
```

Reply:
ah yes, that's a nice idea. Let me make sure it's safe!